### DISCUSSION TITLE
Question: Scaling guide/suggested parameters?

### DISCUSSION THREAD

**Discussion by zaptrem** [archinetai/audio-diffusion-pytorch#6] [General]
Hello! I'm in the process of training a model on a top-40s dataset using your library. However, I want to experiment with long-term consistency, so I've scaled sample rate/channels accordingly to fit ~90s windows during training. I think my results could be improved by further scaling up the number of model parameters, but I'm not sure what to change and by what ratios to get the most bang for my buck/VRAM/compute. Do you guys have a "scaled" config you could share or a general guide (e.g., 2X attention heads, 1.5X mults) for this? Thanks!

**Comment by flavioschneider** [comment]
I'm not an expert on this, and I think it would be hard to tell without running experiments. If I had to guess (seeing how similar models are scaled in for image generation), I would increase the number of resnet blocks `num_blocks: [2, 2, 2, 3, 3, 3]` or `num_blocks: [2, 2, 2, 4, 4, 4]` depending on how large you want to go. You could also play with multipliers to increase the number of channels, e.g. `multipliers: [1, 2, 4, 4, 4, 8, 8]`.

**Comment by zaptrem** [comment]
Thanks! I'll try those settings. So you'd leave the attention features/heads/etc the same?

**Comment by flavioschneider** [comment]
Heads definitely, you could increase `attention_features` to 128 then you would have a total of `128*8=1024` hidden features, which matches Imagen if I'm not wrong. They use twice the number of attention hidden features as channels (since if you have 128 channels with a multiplier of 4 that would be 512 channels).

Btw let me know if you get good results and which setting you end up using :)

**Comment by zaptrem** [comment]
I stuck with the numbers you described in your first comment and left `attention_features` untouched. You can follow the results here: https://wandb.ai/zaptrem/diffusion-pop-3?workspace=user-zaptrem

I also ran into a couple issues with my PC overheating (really hot this weekend!) and doubled the dataset size half way through, which explains the weird loss curves. Additionally, this doesn't include your more recent context_channels commit. Do you think it's worth resetting with the increased attention and context channels? Or seeing this one further through?

Also, does this library use the VAE encoding trick Stable Diffusion uses to increase efficiency?

**Comment by flavioschneider** [comment]
Thanks for sharing! The context channels commit is for some experiments I'm doing with conditioning so it's not necessary for unconditional generation. Hard to tell what's worth trying, I would wait for this experiment to be done and maybe run another where you only change the attention size to compare which is more influent. 

I tried the VAE to increase efficiency but it's very hard to train a good VAE. There's no good loss function for audio, and it's also hard to make diffusion work with that. I would leave that away for now if you don't what to do lots of experiments :)

**Comment by zaptrem** [comment]
> it's very hard to train a good VAE

Could one just steal the pretrained VQVAEs from OpenAI's Jukebox? Or is that type not useful for efficiency improvements like that of Stable Diffusion?

**Comment by flavioschneider** [comment]
I'm not sure that would work since in order to add noise to the encoded input it needs to be in the range -1,1 with a mean of 0. Maybe if properly regularized

**Comment by zaptrem** [comment]
That makes sense. Is there any reason a pyramid of diffusers a-la Jukebox's transformer priors couldn't do a similar job? That was my plan once I got something resembling acceptable results out of this level. Also, is there a rule of thumb for when to end training? Or do people just wait until changes are no longer audible/visible?

**Comment by zaptrem** [comment]
I switched to the larger attention features version and am getting slightly more encouraging results: https://wandb.ai/zaptrem/diffusion-pop-4?workspace=user-zaptrem

I think I should keep scaling.

Is the learning rate falloff determined by number of epochs, or steps?

**Comment by flavioschneider** [comment]
> That makes sense. Is there any reason a pyramid of diffusers a-la Jukebox's transformer priors couldn't do a similar job? 

When you say a pyramid of diffusers do you mean like: a first diffusion model predicting a source at 12kHz, then a second upsampling that to 24kHz, and a third to 48kHz? 

> Also, is there a rule of thumb for when to end training? Or do people just wait until changes are no longer audible/visible?

There isn't. I've noticed that sometimes, even if the loss seems to converge, the quality continues to improve a bit after that. It's hard to find a rule that always applies, since there's no good metric for audio quality. 

> I switched to the larger attention features version and am getting slightly more encouraging results: https://wandb.ai/zaptrem/diffusion-pop-4?workspace=user-zaptrem

That's very interesting! (For some reason, the provided link seems to be dead)

> Is the learning rate falloff determined by number of epochs, or steps?

I didn't add any LR scheduler, but other people I think use InverseLR, CosineAnnealingLR, or LambdaLR scheduling. Also, ideally you would keep a second model with EMA from which you do the sampling so that it's more stable. See for example the trainer in [imagen-pytorch](https://github.com/lucidrains/imagen-pytorch/blob/main/imagen_pytorch/trainer.py). It's something I might add to the trainer in the future. 

Btw, I'm going to move this issue into the general discussion :)

**Reply by zaptrem** [reply]
Fixed the link. I got to almost 250k steps and the results are clear consistent audio. However, where I was expecting temporal progression (e.g., chord progressions, chorus/verse patterns) I usually saw only long riffs/choruses with zero variety. Could that be due to lack of time embeddings or guided diffusion? I also messed up the checkpointing settings so lost the last day of training. However, the recording results are available at the link.


> Learning rate scheduler

There is a setting for it in the .yaml configs like this: "learning_rate: 1e-4"

**Reply by btrude** [reply]
How big is the dataset you are using?

**Reply by zaptrem** [reply]
> How big is the dataset you are using?

About 40-50 hours of modern top-40s music.

**Reply by zaptrem** [reply]
How would you suggest I keep scaling the params for the next try?

**Reply by btrude** [reply]
I'm asking myself the same question 😅 . I have a dataset with 600 hours and my results are worse than yours. Did you change any parameters other than the attention_features?

**Reply by flavioschneider** [reply]
Currently, the model works quite well on monophonic data (e.g. voice, drums, or classical music) but struggles when training on a dataset with many overlapped sources as happens in random music. Guidance might help to increase variety, but I'm not sure that it would fix the frequent choruses with overlapped sounds. 

Scaling might improve this as the model gets more powerful and able to recognize the different sources as separate. Note however that the current model you are using has around 300M params, and is trained for only 250k steps, so it's probably very under trained. I would stop scaling and try to train for longer. 

> There is a setting for it in the .yaml configs like this: "learning_rate: 1e-4" 

That's the learning rate, not the scheduler. The scheduler is not yet implemented (definitely something to consider if you want to properly train for longer).

**Reply by zaptrem** [reply]
>  so it's probably very under trained.

Is there a good way to ballpark params to training steps needed to get a good idea of capabilities? I bit the bullet and started training from the checkpoint I had again. Also, my guess was that 27gb of training data would be a reasonable amount for 1.5gb of model. Should I have gathered more? I assumed I'd need to eventually scale to billions of parameters in order to get results comparable to Jukebox.

> When you say a pyramid of diffusers do you mean like: a first diffusion model predicting a source at 12kHz, then a second upsampling that to 24kHz, and a third to 48kHz?

Exactly. I could also take the raw waveform output of this model, convert it to Jukebox Level 0 VQVAE tokens, then use their level 1 and 0 priors to upscale it. Jukebox L0 transformer sounds great... until you want more than 12-24 seconds, then it forgets what's going on due to the limited context window. The big missing puzzle piece IMO is musicality and long-range musical structure (that's why I'm training it on ~90-second samples).

**Reply by flavioschneider** [reply]
> Is there a good way to ballpark params to training steps needed to get a good idea of capabilities? 

Imagen is trained for 2.5M steps at all stages (400M, 600M, 2B params) with batch sizes of 2048, on 256 TPUs. I don't think that if your dataset is small, the quality will be bad, but will probably overfit. In that case, the model will copy part of songs instead of generating them. The dataset size in Imagen is 460M text/image pairs, if you are training on 1 min chunks that would be equivalent to 320k hours haha, but I'm not sure that it's comparable in this way. 

If you want to keep scaling, maybe try with the attention features again, since you mentioned that the improvement was more noticeable, but first I would try to train for at least 1M steps to see how much the model improves. I've listened to a few of the current samples on your last experiment, and the quality seems already much better.

**Reply by zaptrem** [reply]
Thanks! I'll train to 1M steps then grow the attention features again (is there a ratio they should be increased by to avoid breaking something?) However, I just got our gigabit connection fixed on the training PC (it was previously stuck on an old 802.11N USB stick) so can now download an order of magnitude more training data. Would growing the training set mid-run hurt anything or only help?

**Reply by zaptrem** [reply]
Update: It finally generated something resembling a chorus-verse structure. 362m parameters and only a few days of training on a single 3090. When this architecture gets guidance and a proper scaled version I think it's going to be more impactful than DALL-E 2 and Stable Diffusion combined.

https://user-images.githubusercontent.com/1612230/188251456-adc196d2-40cf-477f-97cc-1740eb3d8587.mp4


What does the use_learned_time_embedding setting do?

**Reply by flavioschneider** [reply]
Really nice! Is this only by training longer, right? 

I agree it's going to be super impactful, now I'm training the new upsampler from 6 to 42kHz and seems to be working very well, almost perfectly. Then we could build generative diffusion models at 6kHz and later upsample that. It's likely that the upsampler could even go lower than 6kHz. You can check it out here (only the voice one has been trained properly for now) https://wandb.ai/schneider/audio/reports/Audio-Diffusion-Echo---VmlldzoyNTU2NTcw?accessToken=sthdn25n8is30gjo2x0w4fs9hwbua23rlbg7o4bv8h17y47xdtruiiyb33aoc5h4 

> What does the use_learned_time_embedding setting do?

Standard discrete diffusion used a fixed time embedding for integer timesteps (time in the sense of the noise conditioning for diffusion) instead of a learned embedding from fractional values as in continuous time diffusion. That flag was used to switch between the two, however I've now removed and fixed it to true, since we only use continuous time (it's superior).

**Reply by zaptrem** [reply]
> Really nice! Is this only by training longer, right?
> 
> I agree it's going to be super impactful, now I'm training the new upsampler from 6 to 42kHz and seems to be working very well, almost perfectly. Then we could build generative diffusion models at 6kHz and later upsample that. It's likely that the upsampler could even go lower than 6kHz. You can check it out here (only the voice one has been trained properly for now) https://wandb.ai/schneider/audio/reports/Audio-Diffusion-Echo---VmlldzoyNTU2NTcw?accessToken=sthdn25n8is30gjo2x0w4fs9hwbua23rlbg7o4bv8h17y47xdtruiiyb33aoc5h4
> 
> > What does the use_learned_time_embedding setting do?
> 
> Standard discrete diffusion used a fixed time embedding for integer timesteps (time in the sense of the noise conditioning for diffusion) instead of a learned embedding from fractional values as in continuous time diffusion. That flag was used to switch between the two, however I've now removed and fixed it to true, since we only use continuous time (it's superior).

Yes, just training longer. Here are some recent results: 

https://user-images.githubusercontent.com/1612230/189036220-1ff7e98a-1a21-4920-8c42-350aae389093.mp4


https://user-images.githubusercontent.com/1612230/189035310-ed653cb3-5408-4916-9356-7b85ba59de1e.mp4


I noticed that this model seems to have a tough time maintaining a consistent beat. Now that this model is nearing 1m steps, I'm hoping to start training a 6khz model with >2X the parameters and ~1.5 orders of magnitude more extremely high-quality training data. Can you suggest any architecture specs to change to make the best use of those extra parameters?

Also, it seems like increasing the size of each epoch (even while leaving batch size alone) increases both VRAM and RAM usage to the point where the training session crashes on my machine. Is there a way to mitigate this?

**Reply by flavioschneider** [reply]
Try with 192 or 256 base `channels`, depending on how big you want to go. The epoch thing is very weird, do you observe a memory leak during training in the GPU VRAM? 

Could you please paste here the parameters you are using as follows (so that it's easier for me and everyone else to keep track)
```yaml
in_channels: ______
channels:  ______
patch_size:  ______
multipliers: ______
factors: ______
num_blocks: ______
attentions:  ______
attention_heads:  ______
attention_features:  ______
attention_multiplier:  ______

total_params: ______
```

**Reply by zaptrem** [reply]
> Try with 192 or 256 base `channels`, depending on how big you want to go. The epoch thing is very weird, do you observe a memory leak during training in the GPU VRAM?
> 
> Could you please paste here the parameters you are using as follows (so that it's easier for me and everyone else to keep track)
> 
> ```yaml
> in_channels: ______
> channels:  ______
> patch_size:  ______
> multipliers: ______
> factors: ______
> num_blocks: ______
> attentions:  ______
> attention_heads:  ______
> attention_features:  ______
> attention_multiplier:  ______
> 
> total_params: ______
> ```

I just uploaded all of the configs + the latest checkpoints for the previous models I trained here (**edit**: fixed link): https://drive.google.com/drive/folders/12yE_eCEnNF_FYYysJqWEKiiU7dktDx3o?usp=sharing

Here are the WIP settings for the 6khz single-channel one I want to train now (I bumped all of the parameters I increased before, but haven't tried base channels yet. Is base channels like a multiplier the entire model (including attention)?):

```YAML
  in_channels: 1
  channels: 128
  patch_size: 16
  resnet_groups: 8
  kernel_multiplier_downsample: 2
  kernel_sizes_init: [1, 3, 7]
  multipliers: [1, 2, 4, 8, 8, 8, 8]
  factors: [4, 4, 4, 2, 2, 2]
  num_blocks: [2, 4, 4, 4, 8, 8]
  attentions: [False, False, False, True, True, True]
  attention_heads: 8
  attention_features: 128
  attention_multiplier: 16
```

However, while updating and using `mono: True` got rid of the error I was running into with single-channel, I didn't save any VRAM in training vs stereo (I assumed the results would be the same as cutting sample length in half). Is that expected?

**Reply by flavioschneider** [reply]
Yeah, reducing the sample length in half reduces by half the number of convolutions and attention tokens. Adding another channel at the beginning has almost no effect, since it's patched down and always expanded to `channels` channels.

**Reply by zaptrem** [reply]
> 

Good to know, thanks. I'm now training with this set of parameters [here](https://wandb.ai/zaptrem/diffusion-pop-5?workspace=user-zaptrem):

```YAML
  in_channels: 2  
  channels: 128
  patch_size: 16
  resnet_groups: 8
  kernel_multiplier_downsample: 2
  kernel_sizes_init: [1, 3, 7]
  multipliers: [1, 2, 4, 8, 8, 8, 8]
  factors: [4, 4, 4, 2, 2, 2]
  num_blocks: [2, 4, 4, 4, 8, 8]
  attentions: [False, False, False, True, True, True]
  attention_heads: 8
  attention_features: 128
  attention_multiplier: 16
```

It seems to be converging much more slowly than the previous models (potentially due to more parameters or me enabling batch accumulation in order to reach a real batch size of 48?) Does anything stick out as suboptimal that I can change before I get too deep into the training run?

**Reply by Karma-Cat** [reply]
@flavioschneider Hello guys! 
First and foremost, I would like to express my gratitude for your exceptional work and fascinating research. I have trained the model on 16kHz mono WAV files, using the following basic parameters for the DiffusionModel:

DiffusionModel(
  net_t=UNetV0, // diffusion (U-Net V0 in this case)
  in_channels=1,
  channels=[8, 32, 64, 128, 256, 512, 512, 1024, 1024],
  factors=[1, 4, 4, 4, 2, 2, 2, 2, 2],
  items=[1, 2, 2, 2, 2, 2, 2, 4, 4],
  attentions=[0, 0, 0, 0, 0, 1, 1, 1, 1],
  attention_heads=8,
  attention_features=64,
  diffusion_t=VDiffusion,
  sampler_t=VSampler
)

I would appreciate some clarification regarding the attention mechanisms. Specifically, are the attention_heads analogous to quantizers, and are the attention_features similar to the size of the vector quantization? Or is my understanding incorrect?

I'm trying to find ways to improve the quality of the synthesized audio(without improving quality of training data)

Thank you in advance!
